File: Python Fundamentals/7. Dictionary/legendary_farming.py
def print_result(final_dict):
    for key, value in final_dict.items():
        print(f"{key}: {value}")



key_materials = {'shards': 0, 'fragments': 0, 'motes': 0}
junk = {}
legendary = {'shards': 'Shadowmourne', 'fragments': 'Valanyr', 'motes': 'Dragonwrath'}
txt = input().split()
winner = False


while not winner:
    for i in range(0, len(txt), 2):
        key = txt[i + 1].lower()
        quantity = int(txt[i])
        if key in key_materials:
            key_materials[key] += quantity
            if key_materials[key] >= 250:
                winner = True
                print(f"{legendary[key]} obtained!")
                key_materials[key] -= 250
                break
        elif key not in junk:
            junk[key] = quantity
        else:
            junk[key] += quantity
    if winner:
        break
    txt = input().split()

key_materials_sorted = dict(sorted(key_materials.items(), key=lambda x: (-x[1], x[0])))
junk_sorted = dict(sorted(junk.items(), key=lambda x: x[0]))

print_result(key_materials_sorted)
print_result(junk_sorted)

File: JavaScript Fundamentals/JavaScript Fundamentals Associative Arrays/legendaryFarming.js
function solve(arr) {
    const needQuantity = 250;
    let legendaryArr = ["shards", "fragments", "motes"];
    let keyMaterials = {
        shards: 0,
        fragments: 0,
        motes: 0
    };
    let otherMaterials = {}
    
    arr = arr.split(' ');

    for (let i = 0; i < arr.length; i += 2) {
        let quantity = Number(arr[i]);
        let material = arr[i + 1].toLowerCase();

        if (legendaryArr.includes(material)) {
            keyMaterials[material] += quantity;
            

            if (keyMaterials[material] >= needQuantity) {
                switch(material) {
                    case 'shards': console.log("Shadowmourne obtained!"); break;
                    case 'fragments': console.log("Valanyr obtained!"); break;
                    case 'motes': console.log("Dragonwrath obtained!"); break;
                }
                keyMaterials[material] -= needQuantity;
                break;
            }
        } else {
            if (otherMaterials.hasOwnProperty(material)) {
                otherMaterials[material] += quantity;
            } else {
                otherMaterials[material] = quantity;
            }
        }
    }

    let sortKeyMaterials = Object.entries(keyMaterials).sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]));
    let sortOtherMaterials = Object.entries(otherMaterials).sort((a, b) => a[0].localeCompare(b[0]));
    
    for (let [key, value] of sortKeyMaterials) {
        console.log(`${key}: ${value}`);
    }

    for (let [key, value] of sortOtherMaterials) {
        console.log(`${key}: ${value}`);
    }
}

solve('3 Motes 5 stones 5 Shards 6 leathers 255 fragments 7 Shards');

solve('123 silver 6 shards 8 shards 5 motes 9 fangs 75 motes 103 MOTES 8 Shards 86 Motes 7 stones 19 silver');